keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,294 @@
+
+*) This way we are not able to distinguish between a volume plugin / CSI driver not installed on a node or it has been installed and it has no limits.
+  * Predicates modified in this KEP assume that storage provided by **in-tree** volume plugin is available all nodes in the cluster. Other predicate(s) will evaluate `PV.spec.nodeAffinity` and filter out nodes that don't have access to the storage.
+  * For CSI drivers, availability of a CSI driver on a node can be checked in `CSINode.spec`. Its handling is out of scope of this KEP, see non-goals.
+
+CSINode example:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: CSINode
+metadata:
+  name: ip-172-18-4-112.ec2.internal
+spec:
+status:
+  allocatable:
+    # AWS node can attach max. 40 volumes, 1 is reserved for the system
+    - name: ebs.csi.aws.com
+      volumes: 39
+```
+
+* Existing scheduler predicates `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are already deprecated.
+  * If any of them is enabled together with `MaxCSIVolumeCountPred`, the deprecated predicate will do nothing (`MaxCSIVolumeCountPred` does the job of counting both in-tree and CSI volumes).
+  * The deprecated predicates will do any useful work only when `MaxCSIVolumeCountPred` predicate is disabled or `CSINode` object does not have limits for a particular driver.
+  * This way, we save CPU by running only one volume limit predicate during deprecation period.
+
+### New API
+
+CSINode gets `Status` struct with `Allocatable`, holding limit of volumes for each volume plugin and CSI driver that can be scheduled to the node.
+
+```go
+type CSINode struct {
+	...
+
+	// spec is the status of CSINode
+	Status CSINodeStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CSINodeStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeStatus struct {
+	// allocatable is a list of volume limits for each volume plugin and CSI driver on the node.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Allocatable []VolumeLimits `json:"allocatable" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=allocatable"`
+}
+
+// VolumeLimits is map CSI driver name -> maximum count of volumes for the driver on the node.
+// For in-tree volume plugins, name of corresponding CSI driver is used.
+// Value can be either:
+// - Positive integer: that's the volume limit.
+// - Zero: such volume cannot be used on the node.
+// - key is missing in VolumeLimits: there is no volume limit, i.e. any number of volumes can be used on the node.
+type VolumeLimits struct {
+	// name of CSI driver. For in-tree volume plugins, name of corresponding CSI driver is used.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// volumes is maximum number of volumes provided by the CSI driver that can be used by the node
+	Volumes int64 `json:"volumes,omitempty" protobuf:"varint,2,opt,name=volumes"`
```

Comment:
Do we really need int64 here? int32 should be more than enough.

Reply:
fixed